### PR TITLE
[QT-318] Add Vault CI bootstrap scenarios

### PR DESCRIPTION
**Pull request by mikebaum**
As part of migrating the CI to use the newly created `vault_ci` AWS account, a set of Terraform modules have been created to simply bootstrapping the repo to use the account. The module `service-user-iam` has been created to setup the IAM policy/role for the service user account. The module `bootstrap_ci` has been created to manage the resources that are created for the `vault_ci` account.

### CODE REVIEW

File: enos/ci/enos-terraform-cli.hcl [hunk 1]
```diff
@@ -0,0 +1,8 @@
+
```
Comment (manthonygfp) [inline enos/ci/enos-terraform-cli.hcl:1]:
```suggestion
```

File: enos/ci/enos-scenario-ci-bootstrap-workspaces.hcl [hunk 2]
```diff
@@ -0,0 +1,28 @@
+
```
Comment (manthonygfp) [inline enos/ci/enos-scenario-ci-bootstrap-workspaces.hcl:1]:
```suggestion
```

File: enos/ci/enos-scenario-bootstrap-ci.hcl [hunk 3]
```diff
@@ -0,0 +1,43 @@
+
```
Comment (manthonygfp) [inline enos/ci/enos-scenario-bootstrap-ci.hcl:1]:
```suggestion
```

File: enos/ci/enos-providers.hcl [hunk 4]
```diff
@@ -0,0 +1,21 @@
+
```
Comment (manthonygfp) [inline enos/ci/enos-providers.hcl:1]:
```suggestion
```

File: enos/ci/enos-terraform.hcl [hunk 5]
```diff
@@ -0,0 +1,98 @@
+
```
Comment (manthonygfp) [inline enos/ci/enos-terraform.hcl:1]:
```suggestion
```

File: enos/ci/modules/create_bootstrap_workspaces/main.tf [hunk 6]
```diff
@@ -0,0 +1,40 @@
+
```
Comment (manthonygfp) [inline enos/ci/modules/create_bootstrap_workspaces/main.tf:1]:
```suggestion
```

File: enos/ci/modules/create_enos_ci_ssh_key/main.tf [hunk 7]
```diff
@@ -0,0 +1,31 @@
+
```
Comment (manthonygfp) [inline enos/ci/modules/create_enos_ci_ssh_key/main.tf:1]:
```suggestion
```

File: enos/ci/enos-variables.hcl [hunk 8]
```diff
@@ -0,0 +1,28 @@
+variable "terraform_plugin_cache_dir" {
+  description = "The directory to cache Terraform modules and providers"
+  type        = string
+  default     = null
+}
+
+variable "tfc_api_token" {
+  description = "The Terraform Cloud QTI Organization API token."
```
Comment (manthonygfp) [inline enos/ci/enos-variables.hcl:8]:
```suggestion
  description = "The Terraform Cloud QTI Organization API token"
```

This is to keep the use of periods consistent between var descriptions

**Review by manthonygfp — APPROVED** [review]
I added some cosmetic suggestions but this looks solid functionally so I see no reason to block the merge for optional suggestions!

File: enos/ci/enos-scenario-ci-bootstrap-workspaces.hcl (hunk 2 above)
Comment (mikebaum) [reply enos/ci/enos-scenario-ci-bootstrap-workspaces.hcl]: done.

File: enos/README.md [hunk 9]
```diff
@@ -140,3 +140,42 @@ downloads the artifact built by the `build.yml` workflow, unzips it, and sets th
...
+[bootstrap_workspaces](./ci/enos-scenario-ci-bootstrap-workspaces.hcl) have been created to simplify
+the process of bootstrapping the CI environment.
+
+**Supported Regions** - enos scenarios are supported in the following regions: 
+`"us-east-1", "us-east-2", "us-west-1", "us-west-2"`
+
+## Bootstrap Process
+These steps should be followed to bootstrap this repo for enos scenario execution:
+
+1. **Setup the root Workspace** - In Terraform Cloud manually create a workspace in the `hashicorp-qti`
```
Comment (ryancragun) [inline enos/README.md:157]: This assumes people know how to get access to Terraform Cloud. Perhaps we should give more context around what `backend` means and why we need to keep remote state. We should tell people how to get access to this TFC account and where to get help.
```suggestion
1. **Set up the root Workspace** - In Terraform Cloud manually create a workspace in the `hashicorp-qti` 
```

File: enos/README.md [hunk 10]
```diff
@@ -140,3 +140,42 @@ downloads the artifact built by the `build.yml` workflow, unzips it, and sets th
...
+
+## Bootstrap Process
+These steps should be followed to bootstrap this repo for enos scenario execution:
+
+1. **Setup the root Workspace** - In Terraform Cloud manually create a workspace in the `hashicorp-qti` 
+   organization named `vault-ci-enos-bootstrap`. This workspace will be used as the backend for the 
+   `bootstrap_workspaces` scenario. When creating the workspace choose the execution mode `local`.
+
+
+2. Get the enos ci ssh public key from a member of the QT team.
```
Comment (ryancragun) [inline enos/README.md:162]: Why don't we keep the public key here in text? It's public.. there's no security risk there.

Comment (ryancragun) [reply enos/README.md:162]: We should tell people what the private key secret name is and where to get it. If they don't have access to that shared 1Password secret perhaps direct them to the slack channel where they can request it.

File: enos/README.md [hunk 11]
```diff
@@ -140,3 +140,42 @@ downloads the artifact built by the `build.yml` workflow, unzips it, and sets th
...
+In order to execute any of the scenarios in this repository, it is first necessary to bootstrap the 
+CI AWS account with the required supporting AWS resources. At this time, the only resource that is 
+required is an EC2 ssh key pair. The scenarios [bootstrap_ci](./ci/enos-scenario-bootstrap-ci.hcl) and 
+[bootstrap_workspaces](./ci/enos-scenario-ci-bootstrap-workspaces.hcl) have been created to simplify
+the process of bootstrapping the CI environment.
+
+**Supported Regions** - enos scenarios are supported in the following regions: 
+`"us-east-1", "us-east-2", "us-west-1", "us-west-2"`
+
+## Bootstrap Process
```
Comment (ryancragun) [inline enos/README.md:154]: I think we're missing the part where people need to request access to the `vault_ci` account in Doormat.

File: enos/README.md [hunk 12]
```diff
@@ -140,3 +140,42 @@ downloads the artifact built by the `build.yml` workflow, unzips it, and sets th
...
+
+1. **Setup the root Workspace** - In Terraform Cloud manually create a workspace in the `hashicorp-qti` 
+   organization named `vault-ci-enos-bootstrap`. This workspace will be used as the backend for the 
+   `bootstrap_workspaces` scenario. When creating the workspace choose the execution mode `local`.
+
+
+2. Get the enos ci ssh public key from a member of the QT team.
+
+
+3. **Create the CI Bootstrap Workspaces** - Each region will have its own workspace and state, to
```
Comment (ryancragun) [inline enos/README.md:165]: I'm curious why need to have different workspaces per-region. Seems that maybe we should instead use the same workspace but repeat the steps with each region provider.

**Review by ryancragun — CHANGES_REQUESTED** [review]
In general I think this is going in the right direction. I was skeptical of why we'd use Enos instead of pure Terraform for this, but perhaps the credential setup makes it worth it. I do think we need to rethink using the matrix for this because that requires different state backends per variant. I would rather repeat steps and have a single state backend per product line.

File: enos/README.md [hunk 13]
```diff
@@ -165,9 +165,8 @@ state will be managed by Terraform Cloud. Here are the steps to configure the Gi
 - Access to the `hashicorp-qti` organization in Terraform Cloud.
 - Full access to the CI AWS account is required.
 
-**Note:** For help with access to Terraform Cloud and the CI Account, contact the QT team on
-the Slack channel mentioned above. Access to the AWS account can be done via Doormat, at: 
-https://doormat.hashicorp.services/.
+**Note:** For help with access to Terraform Cloud and the CI Account, contact the QT team on Slack 
+(#team-quality). Access to the AWS account can be done via Doormat, at: https://doormat.hashicorp.services/.
```
Comment (ryancragun) [inline enos/README.md:169]: Good stuff. Perhaps we should also include that they need to request access for the `vault_ci` account via Doormat. That is done in via Cloud Access > AWS > Request Account Access.

The idea being that we tell them exactly what to do and they don't have to try and figure out what to do in Doormat.

**Review by ryancragun — APPROVED** [review]
Very nice. Having to bootstrap the service user that we then use for the policy is tough, but ultimately I think this is sanest approach.

File: enos/README.md (hunk 13 above)
Comment (mikebaum) [reply enos/README.md:169]: Done.

**Pull request merged**